util/topgen.py

```diff
@@ -1297,83 +1305,102 @@ def render_template(template_path: str, rendered_path: Path,
                         out_path /
                         f"rtl/autogen/top_{topname}_rnd_cnst_pkg.sv",
-                        gencmd=gencmd)
-
-        # Since SW does not use FuseSoC and instead expects those files always
-        # to be in hw/top_{topname}/sw/autogen, we currently create these files
-        # twice:
-        # - Once under out_path/sw/autogen
-        # - Once under hw/top_{topname}/sw/autogen
-        root_paths = [out_path.resolve(), SRCTREE_TOP]
-        out_paths = [
-            out_path.resolve(),
-            (SRCTREE_TOP / "hw/top_{}/".format(topname)).resolve()
-        ]
-        for idx, path in enumerate(out_paths):
-            # C Header + C File + Clang-format file
-
-            # "clang-format" -> "sw/autogen/.clang-format"
-            cformat_tplpath = TOPGEN_TEMPLATE_PATH / "clang-format"
-            cformat_dir = path / "sw/autogen"
-            cformat_dir.mkdir(parents=True, exist_ok=True)
-            cformat_path = cformat_dir / ".clang-format"
-            cformat_path.write_text(cformat_tplpath.read_text())
-
-            # Save the header macro prefix into `c_helper`
-            rel_header_dir = cformat_dir.relative_to(root_paths[idx])
-            c_helper.header_macro_prefix = (
-                "OPENTITAN_" + str(rel_header_dir).replace("/", "_").upper())
-
-            # "top_{topname}.h.tpl" -> "sw/autogen/top_{topname}.h"
-            cheader_path = cformat_dir / f"top_{topname}.h"
-            render_template(TOPGEN_TEMPLATE_PATH / "toplevel.h.tpl",
-                            cheader_path,
-                            helper=c_helper)
-
-            # Save the relative header path into `c_helper`
-            rel_header_path = cheader_path.relative_to(root_paths[idx])
-            c_helper.header_path = str(rel_header_path)
-
-            # "toplevel.c.tpl" -> "sw/autogen/top_{topname}.c"
-            render_template(TOPGEN_TEMPLATE_PATH / "toplevel.c.tpl",
-                            cformat_dir / f"top_{topname}.c",
-                            helper=c_helper)
-
-            # "toplevel_memory.ld.tpl" -> "sw/autogen/top_{topname}_memory.ld"
-            render_template(TOPGEN_TEMPLATE_PATH / "toplevel_memory.ld.tpl",
-                            cformat_dir / f"top_{topname}_memory.ld",
-                            helper=c_helper)
-
-            # "toplevel_memory.h.tpl" -> "sw/autogen/top_{topname}_memory.h"
-            memory_cheader_path = cformat_dir / f"top_{topname}_memory.h"
-            render_template(TOPGEN_TEMPLATE_PATH / "toplevel_memory.h.tpl",
-                            memory_cheader_path,
-                            helper=c_helper)
+                        gencmd=gencmd_sv)
+
+        for addr_space in topcfg['addr_spaces']:
+            default_addr_space = addr_space.get('default', False)
+
+            addr_space_suffix = ""
+            if not default_addr_space:
+                addr_space_suffix = "_" + addr_space['name']
```

Comment:
Given that this appears in several places, maybe it makes sense to pull it out into a function?

Reply:
Agree.